cdap-client/src/main/java/co/cask/cdap/client/AuthorizationClient.java

```diff
@@ -202,10 +202,12 @@ private HttpResponse executePrivilegeRequest(EntityId entityId, Principal princi
     FeatureDisabledException, UnauthenticatedException, IOException, NotFoundException, UnauthorizedException {
     HttpResponse httpResponse = doExecuteRequest(request, HttpURLConnection.HTTP_NOT_FOUND);
     if (HttpURLConnection.HTTP_NOT_FOUND == httpResponse.getResponseCode()) {
-      if (!listAllRoles().contains(new Role(principal.getName()))) {
-        throw new NotFoundException("Role: " + principal.getName());
+      if (principal.getType().equals(Principal.PrincipalType.ROLE) &&
+        httpResponse.getResponseMessage().equals(String.format("%s not found.", principal.getName()))) {
```

Comment:
See my earlier comment and try to understand that Sentry is role based other auth providers might not be. So you cannot make a generic client based for Role which is implementation specific.

Also, since the principal check is happening in the implementation of Authorizer which an user can write as they want you cannot make assumption on what error message they are throwing. An Authorizer can throw exception with message "Role does not exists" and then this will not work even for Role based authorization providers.


Although, this is hacky but it seems like there is no way  around it unless you return different response code for entity not found and principal not found or else you need someway to check which was not found 

If you really need to figure out that here like this then you should be checking for <Entity>NotFoundException message since that is thrown from the handler and is a code in cdap  whose error message cdap controls.